### ISSUE TITLE
passing in meteor run configurations

### ISSUE THREAD

**Issue by hypogealgaol** [meteorhacks/meteord#35]
If I need to run my meteor app with a settings file (meteor run --settings settings.json) how do I build the app ensuring that when I use docker run it will also run it with that settings.json file?

**Comment by arunoda** [comment]
You can get a JSON string for your settings and put it inside `METEOR_SETTINGS` env variables. That's how to use it in production.

**Comment by hypogealgaol** [comment]
so in my dockerfile I just have a METEOR_SETTINGS line?

**Comment by ndarilek** [comment]
I just hit this too.

Would it be wise to check for the existence of settings.json automatically, then do something like (in run_app.sh):

```
if [ -e settings.json ]; then
  export METEOR_SETTINGS=`cat settings.json`
fi
```

Not syntax-checked at all, and my Bash scripting is weak.

This would automatically incorporate settings.json. It would be up to the user to ensure, somehow, that the correct settings.json is at the correct path.

Unless there's some Dockerfile trick I don't know of, this seems like it'd be otherwise somewhat more complicated.

**Comment by hypogealgaol** [comment]
sorry @arunoda the answer you provided isn't so clear to me. Are you saying when I docker run my build I pass in a -e METEOR_SETTINGS: /path/to/settings.json or that I need to specify it elsewhere? I'm trying to test deployment right now

**Comment by ndarilek** [comment]
You have to pass the JSON content of settings.json in as the environment
variable. That may be somewhat easy on the command line, but from tools
like Compose, I don't know how you'd do it.

Perhaps a better way would be a METEOR_SETTINGS_PATH environment
variable. If the run script finds this, it adds that file's content to
METEOR_SETTINGS.

**Comment by hypogealgaol** [comment]
ok that makes sense. my only question is where to define these things, and whether they need to be configured pre-build or not. So just to confirm

docker run -d \
    -e ROOT_URL=http://yourapp.com \
    -e MONGO_URL=mongodb://url \
    -e METEOR_SETTINGS= { json object }
    -e MONGO_OPLOG_URL=mongodb://oplog_url \
    -p 8080:80 \
    yourname/app 

which you're saying is tedious, so instead I would do

docker run -d \
    -e ROOT_URL=http://yourapp.com \
    -e MONGO_URL=mongodb://url \
    -e METEOR_SETTINGS_PATH=/path/to/settings.json
    -e MONGO_OPLOG_URL=mongodb://oplog_url \
    -p 8080:80 \
    yourname/app 

and the run script knows to add to the METEOR_SETTINGS variable automatically? or do I have to define this somewhere?

**Comment by ndarilek** [comment]
I'm saying it would be cool if METEOR_SETTINGS_PATH worked as I
described. Your Dockerfile would then do something like:

COPY settings.json /app/settings.json

And METEOR_SETTINGS_PATH would then be set to /app/settings.json.

Otherwise your settings look OK. I do everything in docker-compose which
lets me define all those variables more conveniently. mupx does
something similar but doesn't give as much freedom as compose does.
Compose is similar, just without the nice setup tooling, and with the
power to, say, hang a Wordpress install off of /blog if you don't want
your app hogging the whole domain.

**Comment by hypogealgaol** [comment]
The way that I ended up doing it is settings an environment variable with export METEOR_SETTINGS = $(cat settings.json), then passing it into the docker run -d ... METEOR_SETTINGS = $METEOR_SETTINGS. Like what you provided. I'm fine doing this for now...

But isn't there some way to specify this in the dockerfile at build time? I was hoping that I could configure docker run -d to automatically use all of the required environment variables without me having to pass them in as flags or something. Or maybe I just start using docker compose

**Comment by ndarilek** [comment]
It's part of Docker's separation of concerns. You should be able to run
an image anywhere, changing its settings at runtime, and not pack the
config in with the image. Ideally the image should only include sensible
defaults that will work everywhere, and any environment-specific config
should be in your docker-compose.yml or similar.

The way I do this is set sensible ENV variables in my Dockerfile for
configuration that should be set everywhere, then use docker-compose for
the rest.

**Comment by bwhitty** [comment]
For anyone who needs help doing this via a `docker-compose.yml` file, this is how you do it:

``` yaml
environment:
        - 'METEOR_SETTINGS={"someKey": {"another": {"andAnother": "socket"}}}'
```

Note the usage of the `'` character.

**Comment by yanndebelgique** [comment]
In order to use the mentioned solutions you would need to stringify the json file. How did you guys manage this ? $(cat settings.json) does not work for me. It is not properly escaped. it tries to run each line of my settings file as a terminal command.

**Comment by phongsakornp** [comment]
@yanndebelgique 

```
export METEOR_SETTINGS=`(cat settings.son)`
```

```
docker run -d .. -e METEOR_SETTINGS="$METEOR_SETTINGS" ...
```

**Comment by athyuttamre** [comment]
@bwhitty @ndarilek were you guys able to use `env_file` in `docker-compose` to automatically import `settings.json` without stringifying it?

I'm wondering if there's a way to define the pretty JSON as an environment variable in a separate file, and then use the filename with the `env_file` option. I'd rather avoid the extra step of having to stringify the JSON. 

The magic of Compose is that all you need to do is run `docker-compose up`, nothing else.

**Comment by ndarilek** [comment]
I gave up. Currently using Rancher to host Docker-based apps, and it 
lets me cut-and-paste environment variables into the web UI. If I paste 
the contents of settings.json, it handles squashing out newlines and 
escaping for me. Would really like to see an integrated solution to ship 
settings.json, which I inject into the build workspace via CI.

**Comment by athyuttamre** [comment]
Yeah, I ended up writing a Makefile to load the variable first and then run docker.

Here's mine: https://gist.github.com/athyuttamre/d1c04cefaa4a56e4fc1c

**Comment by dcmwong** [comment]
Hi,

Just came across this issue and a little disappointed that there isn't a solution for this in the meteor community.  Looking at the meteor source code wouldn't it be simple enough to load the json file from a path if the environment variable is passed as stated before?  Something like 

```
if (process.env.METEOR_SETTINGS_PATH) {
  var settingsFile = process.env.METEOR_SETTINGS_PATH + '/' + process.env.METEOR_SETTINGS_FILENAME;
fs.readFile(settingsFile, 'utf8', function (err,file) {
  if (err) {
   throw new Error("METEOR_SETTINGS_PATH is not valid");
  }
  process.env.METEOR_SETTINGS = file;
});

if (process.env.METEOR_SETTINGS) {
  try {
    Meteor.settings = JSON.parse(process.env.METEOR_SETTINGS);
  } catch (e) {
    throw new Error("METEOR_SETTINGS are not valid JSON: " + process.env.METEOR_SETTINGS);
  }
}
```

to be added in https://github.com/meteor/meteor/blob/87681c8f166641c6c3e34958032a5a070aa2d11a/packages/meteor/server_environment.js

Would anyone be interested this as a feature to be added to Meteor?

Note: I literally typed that code into this box (I haven't tested it).

**Comment by elidoran** [comment]
Workaround:

1. git clone https://github.com/meteor/meteor.git
2. do git checkout <commit> for the specific commit of the release version you're currently using (look that up at [releases](https://github.com/meteor/meteor/releases))
3. symlink its "packages/meteor" package into your app's "packages" directory as "meteor"
4. alter the file "server_environment.js" like below
5. provide an absolute path to your settings file as METEOR_SETTINGS_FILE

```javascript
if (process.env.METEOR_SETTINGS) {
  try {
    Meteor.settings = JSON.parse(process.env.METEOR_SETTINGS);
  } catch (e) {
    throw new Error("METEOR_SETTINGS are not valid JSON: " + process.env.METEOR_SETTINGS);
  }
} else if (process.env.METEOR_SETTINGS_FILE) {
  try {
    Meteor.settings = JSON.parse(
      Npm.require('fs').readFileSync(process.env.METEOR_SETTINGS_FILE, 'utf8')
    );
  } catch (e) {
    throw new Error("METEOR_SETTINGS_FILE is not a valid JSON file: " + process.env.METEOR_SETTINGS_FILE);
  }
}
```

This worked both in development mode and in a production build.

**Comment by michallohnisky** [comment]
Or you can use [ulexus/meteor](https://github.com/CyCoreSystems/docker-meteor) who [supports it](https://github.com/CyCoreSystems/docker-meteor/blob/a8acdd5efe5af2ff70450e4667920f671093011d/entrypoint.sh#L171).

You can than set it in your `docker-compose.yml`:

```
meteor:
  environment:
   - SETTINGS_FILE=/home/meteor/www/<path/to/your_setting_file>.json
```
